modules/core/include/opencv2/core/mat.inl.hpp

```diff
@@ -1096,6 +1096,12 @@ void Mat::push_back(const Mat_<_Tp>& m)
     push_back((const Mat&)m);
 }
 
+template<> inline
+void Mat::push_back(const MatExpr& expr)
+{
+    push_back((const Mat&)expr);
```

Comment:
when you cast a reference to another reference, it's like you cast pointer to another pointer. Use (Mat)expr instead of (const Mat&)expr. In order to verify it, in the test below make the c 2x2 instead of 2x1 (or use another expression, not the transposition). Now it looks like the function works incorrectly in most cases